src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java

```diff
@@ -0,0 +1,102 @@
+package minijava.ir.optimize;
+
+import firm.BackEdges;
+import firm.Graph;
+import firm.bindings.binding_irnode;
+import firm.nodes.*;
+import java.util.*;
+import org.jooq.lambda.tuple.Tuple2;
+
+public class AlgebraicSimplifier extends BaseOptimizer {
+
+  private List<Tuple2<Node, Node>> nodesToExchange;
+
+  @Override
+  public boolean optimize(Graph graph) {
+    this.graph = graph;
+    hasChanged = false;
+    nodesToExchange = new ArrayList<>();
+    BackEdges.enable(graph);
+    fixedPointIteration();
+    replaceNodes();
+    BackEdges.disable(graph);
+    return nodesToExchange.size() > 0;
+  }
+
+  @Override
+  public void visit(Add node) {
+    Optional<Tuple2<Const, Node>> subNodesOpt = split(node);
+    if (!subNodesOpt.isPresent()) {
+      return;
+    }
+    Tuple2<Const, Node> subNodes = subNodesOpt.get();
+    if (subNodes.v1.getTarval().isNull()) { // x * 0  == 0
+      exchange(node, subNodes.v2);
+    } else if (subNodes.v2.getOpCode() == binding_irnode.ir_opcode.iro_Add) {
+      // (x + 8) + 8 == x + 16
+      Add subAddNode = (Add) subNodes.v2;
+      Optional<Tuple2<Const, Node>> subAddNodesOpt = split(subAddNode);
+      if (subAddNodesOpt.isPresent()) {
+        Node constant =
+            graph.newConst(subNodes.v1.getTarval().add(subAddNodesOpt.get().v1.getTarval()));
+        exchange(node, graph.newAdd(node.getBlock(), subAddNodesOpt.get().v2, constant));
+      }
+    }
+  }
+
+  @Override
+  public void visit(Mul node) {
+    Optional<Tuple2<Const, Node>> subNodesOpt = split(node);
+    if (!subNodesOpt.isPresent()) {
+      return;
+    }
+    Tuple2<Const, Node> subNodes = subNodesOpt.get();
+    if (subNodes.v1.getTarval().isNull()) { // x * 0  == 0
+      exchange(node, subNodes.v1);
```

Comment:
Calling this helper method `exchange` is misleading. My first thought was that you are mixing up analysis and optimization phase...